### PR TITLE
DBZ-2862 Switch tracing to OpenTelemetry

### PR DESCRIPTION
**Pull request by vidourem**
- Rewrite SMT with Otel API
- add Interceptor for e2e tracing
- add Otel TextMap getter and setter

### CODE REVIEW

**Comment by Naros** [comment]
@jpechane I wonder if we should consider migrating the Quarkus extension to OpenTelemetry as a part of this PR, so we remain consistent across the portfolio of what tracing framework we're using?

**Comment by jpechane** [comment]
@Naros Thanks for raising that point! Yes you are right we should migrate both sides at the same time.

**Comment by vidourem** [comment]
@jpechane I pushed the adaptation with OpenTelemetry dependencies in the provided scope. I'm not sure if I followed the git workflow correctly.

**Comment by Naros** [comment]
Hi @vidourem it would appear the PR has quite a number of older commits.  Could you rebase your branch on main so that the only commits that are present are your code changes related to DBZ-2862?

**Comment by vidourem** [comment]
@Naros 
Thanks, that was it. 

I chose to extract the interceptor enabling end-to-end traceability in a new module. Seems like a good architectural choice to me. I understand if this is not your opinion, I can make changes if you wish.

**Comment by 1Mark** [comment]
Any progress on this?

**Review by jpechane — COMMENTED** [review]
@vidourem Good job! I am not sure if we add it to 2.2 as is already a bit late. Also before mergeing as this is non-trivial change, could you please prepare

- [ ] a separate PR wit documentation upgrade that will be reviewd by @roldanbob 
- [ ] a PR to the examples repo with example deployment
- [ ] a PR to the `container-images` that would allow enabling the otel support via env var - like ther optional features

File: debezium-core/src/main/java/io/debezium/transforms/tracing/ActivateTracingSpan.java
Comment (Delorien84) [inline debezium-core/src/main/java/io/debezium/transforms/tracing/ActivateTracingSpan.java:122]: Hello, why it is needed to create new instance of `ActivateTracingSpanDelegate` for every message? It is performance penalty (mainly GC stuff) for no obvious reason. `ActivateTracingSpanDelegate` does not save its state or anything like that.

File: debezium-interceptor/pom.xml
Comment (vidourem) [reply debezium-interceptor/pom.xml]: You are right i removed the OpenTelemetry dependencies from the interceptor package.

File: pom.xml
Comment (vidourem) [reply pom.xml]: Done

File: debezium-bom/pom.xml
Comment (vidourem) [reply debezium-bom/pom.xml]: Done

Comment (vidourem) [reply debezium-bom/pom.xml]: Done

File: debezium-bom/pom.xml
Comment (vidourem) [reply debezium-bom/pom.xml]: Done

File: debezium-core/pom.xml
Comment (vidourem) [reply debezium-core/pom.xml]: Done

File: debezium-core/src/main/java/io/debezium/transforms/tracing/ActivateTracingSpan.java
Comment (vidourem) [reply debezium-core/src/main/java/io/debezium/transforms/tracing/ActivateTracingSpan.java]: You are right. I changed the instance method to static method. I took the opportunity to make other minor performance improvements.

**Comment by Naros** [comment]
Hi, @vidourem where do we stand on this one?

**Comment by vjuranek** [comment]
> @vidourem Good job! I am not sure if we add it to 2.2 as is already a bit late. Also before mergeing as this is non-trivial change, could you please prepare
> 
>     * [ ]  a separate PR wit documentation upgrade that will be reviewd by @roldanbob
> 
>     * [ ]  a PR to the examples repo with example deployment
> 
>     * [ ]  a PR to the `container-images` that would allow enabling the otel support via env var - like ther optional features

hi @vidourem besides couple of thing which needs to be done (mentioned by @jpechane in comment above), there's one more: could you please switch to OpenTelemetry also in [debezium quarkus outbox extension](https://github.com/debezium/debezium/blob/main/debezium-quarkus-outbox/runtime/src/main/java/io/debezium/outbox/quarkus/internal/DebeziumTracerEventDispatcher.java) as well as its [reactive variant](https://github.com/debezium/debezium/blob/main/debezium-quarkus-outbox-reactive/runtime/src/main/java/io/debezium/outbox/reactive/quarkus/internal/DebeziumTracerEventDispatcher.java)? I hope this should be complete list now :-)

**Comment by vidourem** [comment]
Hi all, I've been very busy the last few days with another priority project for me. I think I'll be able to get back on that feature soon.

**Comment by vidourem** [comment]
@jpechane 

Hello,

Concerning documentation, it is stated in the CONTRIBUTE.md

_"Any documentation updates should be part of the pull request you submit for the code change."_

Do I still need to make a separate PR?

**Comment by vjuranek** [comment]
hi @vidourem , first, thanks for the continuation of the work! 

For small doc changes it's fine to include them into the same PR, but for bigger changes (like this on, which is expected to be more than a few lines long) it's better to put it into separate PR as the documentation will be reviewed by @roldanbob  (i.e. there will be different reviewers, can be different speed of views etc. so it makes sense to split it in this case).

Speaking about the review speed - I'll be offline whole next week (and @jpechane  as well), so unless someone else reviews it, please apologize the delay with the review. We will review it week after next week. Thanks for understanding.

**Comment by vidourem** [comment]
Hello,
I've decided to develop the outbox example at the same time as writing the documentation. This will guide me in writing the documentation.

I'm progressing well for the moment. 

However, I've had a few problems with Strimzi. The latest version of Strimzi provides OpenTelemetry dependencies in version 1.19.0. This poses a compatibility problem with version 1.23.0, which I have chosen to use in Debezium. In fact, a class that I use in the interceptor has been renamed. So I didn't enable traceability with the strimzi option and I added the OpenTelemetry agent and dependencies to the classpath by modifying the Dockerfile in the example.

Does anyone have the possibility of asking the Strimzi team to take OpenTelemetry 1.23.0+ into account in the next version, ideally the latest version 1.27.0?

I also need to get the posgres connector in Snapshot version to build the example docker image. Have the changes I've made been embedded and where can I get the package? For the moment I'm retrieving the connector package locally.

Finally, for the interceptor, I also need to retrieve the jar to include it in the Strimzi classpath. I may need to publish the package to allow users to retrieve it, as it is not included in the connector packages.

I hope to finish a first iteration this week.

David.

**Comment by vidourem** [comment]
Hello,
@jpechane @vjuranek 

In this PR, I've introduced a commit that upgrades the Quarkus module to OpenTelemetry.

I've successfully deployed and tested the outbox example on my personal repository, which can be found here: https://github.com/vidourem/debezium-examples/tree/DBZ-2862/outbox

Originally, I was able to execute the example using Strimzi distribution, but it left me unsatisfied. To enhance this, we would require Strimzi to incorporate OpenTelemetry version 1.23.0 or later.

I was also able to get the example running with the debezium-connect image.

In my commit, I've left the Dockerfile and docker-compose for Strimzi tagged with the .strimzi suffix.

I encountered some issues while trying to configure the database user password in Quarkus dev mode.

An error related to the timestamp configuration in the connector has been addressed by temporarily removing the timestamp config.

Furthermore, I'm experiencing some conflicts with Vertx that seem to originate from OpenTelemetry. My next move is to attempt an OpenTelemetry version upgrade.

To build the debezium-connect image, I would need access to your repository housing the modified version of the Postgres connector. Additionally, access to the interceptor jar would be required. I also need to add OpenTelemetry dependencies to the classpath, which I am currently sourcing from Maven Central. Should I retrieve these from another repository?

Docker image creation is not exactly my area of expertise. Any assistance with this would be greatly appreciated.

Best,
David

**Comment by vidourem** [comment]
Hello,
@jpechane @vjuranek
I started modifying the image as best I could.
https://github.com/vidourem/container-images/tree/DBZ-2862

**Review by vjuranek — COMMENTED** [review]
hi @vidourem, thanks a lot for pushing this PR forward! I skimmed through the changes and overall LGTM, so far I have only one question about java bytecode target level.

As for your questions - have you tried to send a PR for Strmzi for bumping OTEL version? This seems to be the most straightforward way. IMHO we should keep the version in sync.
As for the libs - they should be probably included directly in the images so that users can use tracing without any further burden. However, in both cases I'd defer to @jpechane for final decision.

I'm sorry, I'll be next week again offline, but once I return, I'll do more detail review and testing.

**Comment by vidourem** [comment]
Hi @vjuranek, @jpechane 

Following the insightful recommendations from the Strimzi team https://github.com/orgs/strimzi/discussions/8865, I've implemented a dynamic solution to manage changes in OpenTelemetry alpha versions. The approach uses introspection at runtime to detect which class is present in the classpath, and then delegates the logic to the correct instances of OpenTelemetry's interceptor accordingly. 

This solution ensures compatibility with both Strimzi, which uses OpenTelemetry versions earlier than 1.23.0, and with other Kafka Connect distributions that may use OpenTelemetry version 1.23.0 or later.

I've also eliminated the use of OpenTelemetry's 'ConfigUtils' as it's part of the non-public API and may not always be available in different versions.

I hope this approach is in line with our objectives, and I look forward to your feedback.

Best,

**Comment by vjuranek** [comment]
hi @vidourem , this approach seems completely fine for me - thanks for IMHO the most universal solution. No matter which version you'd use, sooner or later there will be request in the community to support the other version:-). I'd just appreciate short comments in `DebeziumTracingProducerInterceptor` constuctor and `InterceptorVersion` to make clear why we need to use dynamic loading, as at some point, once OTEL API is stabilized, we will remove it and comments can give quick hint what can be removed (and what is clear now can be not that clear some time later).

File: debezium-interceptor/src/main/java/io/debezium/tracing/DebeziumTracingProducerInterceptor.java
Comment (vjuranek) [inline debezium-interceptor/src/main/java/io/debezium/tracing/DebeziumTracingProducerInterceptor.java:40]: IMHO this deserves comment

File: debezium-interceptor/src/main/java/io/debezium/tracing/DebeziumTracingProducerInterceptor.java
Comment (vjuranek) [inline debezium-interceptor/src/main/java/io/debezium/tracing/DebeziumTracingProducerInterceptor.java:55]: should be enough only to check is `onSendMethod` is `null`

File: debezium-interceptor/src/main/java/io/debezium/tracing/InterceptorVersion.java
Comment (vjuranek) [inline debezium-interceptor/src/main/java/io/debezium/tracing/InterceptorVersion.java:40]: as we iterate over two classes which typically only one will be on the class path we may eventually always get error in the log if the first one won't be present and at the same time this won't be a real issue as second class would be instantiated. I'd log it on the debug level.

File: debezium-interceptor/src/main/java/io/debezium/tracing/InterceptorVersion.java
Comment (vjuranek) [inline debezium-interceptor/src/main/java/io/debezium/tracing/InterceptorVersion.java:53]: Possibly same here. The error is eventually logged in `DebeziumTracingProducerInterceptor`.

**Review by vjuranek — CHANGES_REQUESTED** [review]
Reviewed (not tested it yet) and overall LGTM, I have only few minor comments. Great work @vidourem !

File: debezium-core/src/main/java/io/debezium/transforms/tracing/ActivateTracingSpan.java
Comment (vidourem) [reply debezium-core/src/main/java/io/debezium/transforms/tracing/ActivateTracingSpan.java:127]: Currently, the check is performed externally by the EventRouterDelegate class. I didn't change this intentionally in order to concentrate on the change of tracing library, I also wondered about this external check. That said, it seems to follow a certain logic. Now that the library change is fairly stable, I can try a refactoring to move the check to the ActivateTracingSpan class, or simply add a check to the ActivateTracingSpan class, which would mean doing the same check twice. It's up to you to tell me what you want.

**Comment by vjuranek** [comment]
@vidourem I finished the testing and everything looks good. As I wrote above, great work!
Will you send a PR with updating our [outbox example](https://github.com/debezium/debezium-examples/tree/main/outbox) or I should handle that (I saw you did most/all the work so I guess you will send it)?
Another thing which needs to be done is updating the documentation.
Thanks!

**Comment by jpechane** [comment]
The documentation can be sent as a separate PR so @roldanbob tak take a look at it.

**Comment by vidourem** [comment]
@vjuranek 
>  Will you send a PR with updating our [outbox example](https://github.com/debezium/debezium-examples/tree/main/outbox) or I should handle that (I saw you did most/all the work so I guess you will send it)? 

Of course, you will have to terminate the debezium-strimzi Dockerfile when the packages are available. For now, you have to package the postgres connector and the interceptor locally.

https://github.com/debezium/debezium-examples/pull/327

**Comment by vjuranek** [comment]
@vidourem are you going to address those minor comments?
@jpechane from my point of view looks good now (those minor comments I can do myself before merge if not addressed by @vidourem). We need to include new bits (kafka interceptor) into container images, but this can be done once it's release (same applies to outbox example update).

**Comment by vidourem** [comment]
Hi @vjuranek,

I've made 4 changes out of the 5 comments. I asked you a question about the first comment because I'm not sure what to do. 

https://github.com/debezium/debezium/commit/ff5d3e391dbf8e5cafa060617506973b353c3603

https://github.com/debezium/debezium/pull/4270#discussion_r1279231459

**Comment by jpechane** [comment]
@vjuranek No comments from my side. Would it be possible to prepare container image changes in advance?

**Comment by vidourem** [comment]
@jpechane , @vjuranek 
I made a PR for container image, it is not my area of expertise, I am interested in your remarks

https://github.com/debezium/container-images/pull/343

File: debezium-core/src/main/java/io/debezium/transforms/tracing/ActivateTracingSpan.java
Comment (vjuranek) [reply debezium-core/src/main/java/io/debezium/transforms/tracing/ActivateTracingSpan.java:127]: Make sense, we can do some more refactoring later.

**Review by vjuranek — APPROVED** [review]
@vidourem LGTM now. Thanks a lot!

**Comment by vjuranek** [comment]
Merged, great work - thanks a lot @vidourem !

**Comment by vjuranek** [comment]
@vidourem one more thing - are you going to send a PR with documentation update? Thanks!

**Comment by vidourem** [comment]
@vjuranek 

Yes, I'll try to finish before I go on holiday.

**Comment by vjuranek** [comment]
@vidourem thanks, looking forward to it

**Comment by vidourem** [comment]
The PR for documentation https://github.com/debezium/debezium/pull/4774

**Pull request closed without merging**
